### PR TITLE
Feature evaluate progress

### PR DESCRIPTION
**Pull request by bvanessen**

### CODE REVIEW

File: include/lbann/callbacks/evaluate_progress.hpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ *  There are many algorithmic variations to be explored:
+ *    - How is data is divvied up amongst the trainers. Is it strictly
+ *      partitioned, partially shared, or completely replicated?
+ *    - What model components are exchanged? Just the trainable weights,
+ *      or a subset of the weights? Hyperparameters?
+ *    - Can this be used to explore model architectures?
+ *
+ */
+class evaluate_progress : public callback_base
```
Comment (benson31) [inline include/lbann/callbacks/evaluate_progress.hpp:59]:
```suggestion
class evaluate_progress final : public callback_base
```

File: src/callbacks/evaluate_progress.cpp [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+#include <algorithm>
+#include <numeric>
+#include <set>
+#include <sstream>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+#include <vector>
```
Comment (benson31) [inline src/callbacks/evaluate_progress.cpp:56]: Um.

File: src/callbacks/evaluate_progress.cpp [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  return *this;
+}
+
+void evaluate_progress::on_batch_begin(model* m)
+{
+  auto& local_model = *m;
+  auto& context = local_model.get_execution_context();
+  auto&& comm = *local_model.get_comm();
+
+  // Check whether to start EVALUATE_PROGRESS round
```
Comment (tbennun) [inline src/callbacks/evaluate_progress.cpp:128]:
```suggestion
  // Check whether to evaluate the model
```

**Review by tbennun — APPROVED** [review]

File: src/callbacks/evaluate_progress.cpp (hunk 2 above)
Comment (bvanessen) [reply src/callbacks/evaluate_progress.cpp]: I have reduced this.

**Pull request merged**